storage/remote/queue_manager.go

```diff
@@ -1564,16 +1626,20 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
 			if !ok {
 				return
 			}
-			switch s.qm.rwFormat {
-			case Version1:
-				nPendingSamples, nPendingExemplars, nPendingHistograms := populateTimeSeries(batch, pendingData, s.qm.sendExemplars, s.qm.sendNativeHistograms)
-				n := nPendingSamples + nPendingExemplars + nPendingHistograms
-				s.sendSamples(ctx, pendingData[:n], nPendingSamples, nPendingExemplars, nPendingHistograms, pBuf, &buf)
-			case Version2:
-				nPendingSamples, nPendingExemplars, nPendingHistograms, nPendingMetadata := populateV2TimeSeries(&symbolTable, batch, pendingMinStrData, s.qm.sendExemplars, s.qm.sendNativeHistograms)
-				n := nPendingSamples + nPendingExemplars + nPendingHistograms
-				s.sendV2Samples(ctx, pendingMinStrData[:n], symbolTable.LabelsStrings(), nPendingSamples, nPendingExemplars, nPendingHistograms, nPendingMetadata, &pBufRaw, &buf)
-				symbolTable.clear()
+			// Resend logic on 406
+			// ErrStatusNotAcceptable is a new error defined in client.go
+
+			// Work out what version to send based on the last header seen and the QM's rwFormat setting
+			// TODO(alexg) - see comments below about retry/renegotiate design
+			for attemptNos := 1; attemptNos <= 3; attemptNos++ {
+				lastHeaderSeen := s.qm.storeClient.GetLastRWHeader()
+				compression, rwFormat := negotiateRWProto(s.qm.rwFormat, lastHeaderSeen)
+				sendErr := attemptBatchSend(batch, rwFormat, compression, false)
+				if sendErr == nil || !(errors.Is(sendErr, ErrStatusNotAcceptable) || errors.Is(sendErr, ErrStatusBadRequest)) {
+					// It wasn't an error, or wasn't a 406 or 400 so we can just stop trying
+					break
+				}
+				// If we get either of the two errors (406, 400) we loop and re-negotiate
```

Comment:
And following this same line: `negotiateRWProto` could return an error if parties could not agree on a version, and treat that as unrecoverable.

Reply:
Yeah, there's two kinds of "recoverable":
* A 5xx response is considered "recoverable" in the current Prometheus codebase sense as the retry function will resend the same payload after a short delay, no concept of any repackaging/re-encoding
* A 406 response (which we deem "recoverable" in the spec) may require creating a new payload (e.g. v1 instead of v2 or vice versa, and/or a different encoding) which is beyond what the existing `attemptStore` function does.

This is why I described the 406 as "non-recoverable" as it wasn't "recoverable" in the same way as a 5xx error.

The main problems I found with trying to make use of the same logic as the `attemptSend()` method are:
* it would require passing in a whole load of parameters in order to have the information there to rebuild the payload
* the two payload creation functions take slightly different sets of parameters
so I opted to keep it higher in the stack.

I'll go back and take a look at this bearing in mind the comments above and see what could be done.